database/buildSchema/43_views_functions_triggers.sql

```diff
@@ -1549,12 +1549,332 @@ $$
+            ) 
+        UPDATE public.application
+            SET reviewer_list = (SELECT reviewers FROM lists),
+            assigner_list = (SELECT assigners FROM lists)
+            WHERE id = NEW.application_id;
+    END IF;
+
+    -- Get list of reviewer_ids to update for 
+    IF TG_OP = 'UPDATE' THEN
+        reviewer_ids = ARRAY(
+            SELECT reviewer_id FROM review_assignment
+            WHERE application_id = NEW.application_id
+        );
+    ELSE
+        reviewer_ids = ARRAY[NEW.reviewer_id];
+    END IF;
+
+    FOREACH rev_id IN ARRAY reviewer_ids LOOP
+        -- Remove existing
+        UPDATE public.application_reviewer_action
+            SET reviewer_action = NULL
+            WHERE application_id = NEW.application_id
+            AND user_id = rev_id;
+
+        -- Insert new
+        WITH reviewer_action AS (
+            SELECT reviewer_action
+                FROM single_application_detail(NEW.application_id, rev_id)
+            ) 
+        INSERT INTO public.application_reviewer_action
+            (user_id, application_id, reviewer_action)  
+        VALUES(
+            rev_id,
+            NEW.application_id,
+            (SELECT reviewer_action FROM reviewer_action)
+         )
+         ON CONFLICT (user_id, application_id)
+         DO UPDATE
+            SET reviewer_action = (SELECT reviewer_action FROM reviewer_action);
+    END LOOP;
+    
+    
+    -- ASSIGNERS
+    IF TG_OP = 'UPDATE' THEN
+        assigner_ids = ARRAY(
+            SELECT assigner_id FROM review_assignment
+            WHERE application_id = NEW.application_id
+            AND assigner_id IS NOT NULL
+        );
+    ELSE
+        CASE WHEN NEW.assigner_id IS NULL THEN
+                assigner_ids = ARRAY[];
+            ELSE
+                assigner_ids = ARRAY[NEW.assigner_id];
+            END CASE;
+    END IF;
+
+    FOREACH ass_id IN ARRAY assigner_ids LOOP
+        -- Remove existing
+        UPDATE public.application_reviewer_action
+            SET assigner_action = NULL
+            WHERE application_id = NEW.application_id
+            AND user_id = ass_id;
+
+        -- Insert new
+        WITH assigner_action AS (
+            SELECT assigner_action
+                FROM single_application_detail(NEW.application_id, ass_id)
+            ) 
+        INSERT INTO public.application_reviewer_action
+            (user_id, application_id, assigner_action)  
+        VALUES(
+            ass_id,
+            NEW.application_id,
+            (SELECT assigner_action FROM assigner_action)
+         )
+         ON CONFLICT (user_id, application_id)
+         DO UPDATE
+            SET assigner_action = (SELECT assigner_action FROM assigner_action);
+    END LOOP;
+
+    -- Clean up NULL records
+    DELETE FROM public.application_reviewer_action
+    WHERE application_id = NEW.application_id
+    AND reviewer_action IS NULL
+    AND assigner_action IS NULL;
+
+RETURN NULL;
+
+END;
+$$
+LANGUAGE plpgsql;
```

Comment:
Btw really like the insert to and on conflict, I think it avoids re-building indexes if the fields were otherwise dropped

Reply:
> Another approach is to do calculation for all assigners and users for this application, I did a quick bench it's about 2 seconds on demo4, to calculate actions for all reviewers/assigners for one application, this would be added to all upsert/insert operations on review assignments, hmm.

The problem I had is that the review_assignment records get inserted one by one and so trigger an event for each. So we don't want to update them *all* for *every* insert, hence inserts only update actions for that one assignment record. Whereas an update is done individually, so we can update *all* records associated with that application.

I think with your suggestion, we'd run the whole thing for every assignment being added, which would be very inefficient, not to mention pointless as it would generate an incomplete list for every insert until the last one is added.